ops/testing.py

```diff
@@ -2216,17 +2325,34 @@ def _storage_remove(self, storage_id: str):
         index = int(index)
         self._storage_list[name].pop(index, None)
 
-    def action_get(self):  # type:ignore
-        raise NotImplementedError(self.action_get)  # type:ignore
-
-    def action_set(self, results):  # type:ignore
-        raise NotImplementedError(self.action_set)  # type:ignore
-
-    def action_log(self, message):  # type:ignore
-        raise NotImplementedError(self.action_log)  # type:ignore
-
-    def action_fail(self, message=''):  # type:ignore
-        raise NotImplementedError(self.action_fail)  # type:ignore
+    def action_get(self) -> Dict[str, Any]:
+        action_name = os.environ["JUJU_ACTION_NAME"]
+        params: Dict[str, Any] = {}
+        for name, action_meta in self._meta.actions[action_name].parameters.items():
+            if "default" in action_meta:
+                params[name] = action_meta["default"]
+        params.update(self._operation.parameters)
+        return params
+
+    def action_set(self, results: Dict[str, Any]):
+        for key in "stdout", "stderr", "stdout-encoding", "stderr-encoding":
+            if key in results:
+                # Match Juju's error message.
+                raise model.ModelError(f'ERROR cannot set reserved action key "{key}"')
+        # Although it's not necessary, we go through the same flattening process
+        # as the real backend, in order to give Charmers advance notice if they
+        # are setting results that will not work.
+        # This also does some validation on keys to make sure that they fit the
+        # Juju constraints.
+        model._format_action_result_dict(results)
```

Comment:
Alternatively, I could split `_format_action_result_dict` into validation and formatting (at the cost of looping through the (likely small) dict twice, as well as potentially a few extra function calls since it's done recursively). The comment works for me, so will do that unless you suggest otherwise :)

Fwiw, I originally (and this was somewhat implied in the spec) planned to do the flatten and then do the unflatten that's done on the Juju side, and give back that resulting object, so it would have been used. However, when implementing the unflatten part it seemed a bit pointless to be doing that (but I do feel it's valuable to do the flatten to get any errors there early).